src/java/io/bazel/rulesscala/worker/Worker.java

```diff
@@ -0,0 +1,114 @@
+package io.bazel.rulesscala.worker;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.lang.SecurityManager;
+import java.security.Permission;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.nio.charset.StandardCharsets;
+
+import com.google.devtools.build.lib.worker.WorkerProtocol;
+
+public final class Worker {
+
+    public static interface Interface {
+	public void work(String[] args) throws Exception;
+    }
+
+    public static void workerMain(String workerArgs[], Interface workerInterface) throws Exception {
+	if (workerArgs.length > 0 && workerArgs[0].equals("--persistent_worker")) {
+
+	    System.setSecurityManager(new SecurityManager() {
+		    @Override
+		    public void checkPermission(Permission permission) {
+			Matcher matcher = exitPattern.matcher(permission.getName());
+			if (matcher.find())
+			    throw new ExitTrapped(Integer.parseInt(matcher.group(1)));
+		    }
+		});
+
+	    InputStream stdin = System.in;
+	    PrintStream stdout = System.out;
+	    PrintStream stderr = System.err;
+	    ByteArrayOutputStream outStream = new ByteArrayOutputStream();
+	    PrintStream out = new PrintStream(outStream);
+
+	    System.setIn(new ByteArrayInputStream(new byte[0]));
+	    System.setOut(out);
+	    System.setErr(out);
+
+	    try {
+		while (true) {
+		    WorkerProtocol.WorkRequest request =
+			WorkerProtocol.WorkRequest.parseDelimitedFrom(stdin);
+
+		    int code = 0;
+
+		    try {
+			workerInterface.work(stringListToArray(request.getArgumentsList()));
+		    } catch (ExitTrapped e) {
+			code = e.code;
+		    } catch (Exception e) {
+			System.err.println(e.getMessage());
+			e.printStackTrace();
+			code = 1;
+		    }
+
+		    WorkerProtocol.WorkResponse.newBuilder()
+			.setOutput(outStream.toString())
+			.setExitCode(code)
+			.build()
+			.writeDelimitedTo(stdout);
+
+		    out.flush();
+		    outStream.reset();
+		}
+	    } catch (IOException e) {
+	    } finally {
+		System.setIn(stdin);
+		System.setOut(stdout);
```

Comment:
not blocking this PR but do you think we can somehow get rid of this? `setOut` and `seterr` were deprecated at 2.11.
I tried moving to `withOut` a while back but it broke (don't remember how exactly)

Reply:
Err, I think you've got Scala brain right now instead of Java brain!

These methods aren't deprecated on the [Java System class](https://docs.oracle.com/javase/7/docs/api/java/lang/System.html ) but are deprecated on [Scala's Console object](https://www.scala-lang.org/api/2.12.0/scala/Console$.html). We're in Java land right now, so we should be good!

Reply:
Nice! That’s definitely a win for this PR :)